### PR TITLE
[7.0] npg modules - account_banking_make_deposit, account_banking_reconciliation

### PR DESCRIPTION
**Pull request by bwrsandman**
Porting NPG modules from https://code.launchpad.net/openerp-usa

Many improvements and fixes have been added including multi-company support.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
`npg_` makes reference to any company? You should remove it then.

**Comment by llacroix** [comment]
I might be wrong but travis doesn't seem to be building on commits that I pushed. For example, the commit 48fc0d1 should fail as I didn't push the fix for dangerous default value. Only the last build should pass.

**Comment by bwrsandman** [comment]
I don't think the tests catch a default dict parameter...
It would be good if they did, but the tests don't seem to crash despite the dangerous value.

**Comment by llacroix** [comment]
@bwrsandman The pylint test here fail with dangerous default values. https://travis-ci.org/OCA/bank-payment/jobs/54784917 

```
pylint --rcfile=${PYLINT_CONFIG_DIR}/travis_run_pylint.cfg ${MODULES_TO_TEST}

************* Module bank-payment.account_banking_reconciliation.account_banking_reconciliation

account_banking_reconciliation/account_banking_reconciliation.py:73: [W0102(dangerous-default-value), bank_acc_rec_statement.copy] Dangerous default value {} as argument
```

**Comment by bwrsandman** [comment]
@llacroix please have a look at https://github.com/savoirfairelinux/openerp-usa/issues/1 and see if this patch is relevant and can be integrated.

> Exlcude the journal with type='situation' that will produce an error (**\* already confirm ) when the accounts extend more than one financial year.
> 
> To re-produce the error :
> a. have two financial year
> b. close first financial year
> c. in 2nd financial year, create bank reconciliation, -> refresh() -> error.

File: account_banking_reconciliation/account_banking_reconciliation.py [hunk 1]
```diff
@@ -0,0 +1,1224 @@
...
+                line.move_line_id.id
+                for line in obj.credit_move_line_ids + obj.debit_move_line_ids
+                if line.move_line_id
+            ]
+
+            domain = [
+                ('id', 'not in', move_line_ids),
+                ('account_id', '=', obj.account_id.id),
+                ('move_id.state', '=', 'posted'),
+                ('cleared_bank_account', '=', False),
```
Comment (bwrsandman) [inline account_banking_reconciliation/account_banking_reconciliation.py]: as @fanzalika pointed out, `('journal_id.type', '!=', 'situation'),` should be added

File: account_banking_reconciliation/account_banking_reconciliation.py [hunk 2]
```diff
@@ -0,0 +1,1224 @@
...
+            #    of the selected GL account
+            # 2. Journal items which are not assigned to previous
+            #    bank statements
+            # 3. Date less than or equal to ending date provided the
+            #    'Suppress Ending Date Filter' is not checked
+            # get previous uncleared entries
+            domain = [
+                ('account_id', '=', account_id),
+                ('move_id.state', '=', 'posted'),
+                ('cleared_bank_account', '=', False)
```
Comment (bwrsandman) [inline account_banking_reconciliation/account_banking_reconciliation.py]: as @fanzalika pointed out, `('journal_id.type', '!=', 'situation'),` should be added

**Comment by llacroix** [comment]
@bwrsandman Could you have a look, it's green again

File: account_banking_reconciliation/account_banking_reconciliation.py (hunk 1 above)
Comment (llacroix) [inline account_banking_reconciliation/account_banking_reconciliation.py]: Added the journal_id.type to the domain line below

File: account_banking_make_deposit/account_banking_make_deposit.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
+# -*- coding: utf-8 -*-
```
Comment (bwrsandman) [inline account_banking_make_deposit/account_banking_make_deposit.py]: This file is too long, it should be split up according to model.

File: account_banking_make_deposit/account_banking_make_deposit_view.xml [hunk 4]
```diff
@@ -0,0 +1,162 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bwrsandman) [inline account_banking_make_deposit/account_banking_make_deposit_view.xml]: This only deals with `deposit.ticket` please rename to `views/deposit_ticket.xml` or `deposit_ticket_view.xml`

File: account_banking_make_deposit/demo/account_banking_make_deposit_demo.xml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
```
Comment (bwrsandman) [inline account_banking_make_deposit/demo/account_banking_make_deposit_demo.xml]: This only deals with `deposit.method`, please rename to `demo/deposit_method.xml`

File: account_banking_make_deposit/wizard/add_deposit_items_view.xml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+
+		<!-- Add Deposit Items -->
+		<record id="view_add_deposit_items" model="ir.ui.view">
+            <field name="name">Add Deposit Items</field>
```
Comment (bwrsandman) [inline account_banking_make_deposit/wizard/add_deposit_items_view.xml]: Odd alignment, this should be fixed

File: account_banking_make_deposit/wizard/add_deposit_items_view.xml [hunk 7]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bwrsandman) [inline account_banking_make_deposit/wizard/add_deposit_items_view.xml:1]: Mix of tabs and spaces here, that could be cleaned up with indents of 2 or 4 spaces to be consistent.

File: account_banking_reconciliation/account_banking_reconciliation.py [hunk 8]
```diff
@@ -0,0 +1,1227 @@
...
+            payment_rate = currency_obj._get_conversion_rate(
+                cr, uid, acur, ccur, context=ctx2)
+            val['currency_rate'] = payment_rate
+            val['currency_rate_label'] = self._get_currency_help_label(
+                cr, uid, acur.id, payment_rate, ccur.id, context)
+            val['multi_currency'] = True
+
+        return res
+
+    _name = "bank.acc.rec.statement"
```
Comment (bwrsandman) [inline account_banking_reconciliation/account_banking_reconciliation.py]: You might want to bring this to the top.

File: account_banking_reconciliation/account_banking_reconciliation.py [hunk 9]
```diff
@@ -0,0 +1,1227 @@
...
+            'unique (name, company_id, account_id)',
+            (
+                'The name of the statement must be unique per company '
+                'and G/L account!'
+            )
+        )
+    ]
+
+
+class bank_acc_rec_statement_line(orm.Model):
```
Comment (bwrsandman) [inline account_banking_reconciliation/account_banking_reconciliation.py]: This class should be in it's own file

File: account_banking_reconciliation/account_move_line.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        'bank_acc_rec_statement_id': fields.many2one(
+            'bank.acc.rec.statement',
+            'Bank Acc Rec Statement',
+            help="The Bank Acc Rec Statement linked with the journal item"
+        ),
+        'draft_assigned_to_statement': fields.boolean(
+            'Assigned to Statement? ',
+            help='Check if the move line is assigned to statement lines'
+        )
+    }
```
Comment (bwrsandman) [inline account_banking_reconciliation/account_move_line.py]: `'? '` could be removed from these bools, it's kind of redundant, and I don't see anywhere else in Odoo that labels bools like it.

File: account_banking_reconciliation/i18n/fr.po [hunk 11]
```diff
@@ -0,0 +1,577 @@
+# Translation of OpenERP Server.
```
Comment (bwrsandman) [inline account_banking_reconciliation/i18n/fr.po]: account_banking_reconciliation/i18n/fr.mo should not be commited

File: account_banking_reconciliation/security/ir.model.access.csv [hunk 12]
```diff
@@ -0,0 +1,5 @@
+"id","name","model_id:id","group_id:id","perm_read","perm_write","perm_create","perm_unlink"
+"access_bank_acc_rec_statement_preparer","access.bank.acc.rec.statement.preparer","model_bank_acc_rec_statement","group_bank_stmt_preparer",1,1,1,1
+"access_bank_acc_rec_statement_verifier","access.bank.acc.rec.statement.verifier","model_bank_acc_rec_statement","group_bank_stmt_verifier",1,1,1,1
+"access_bank_acc_rec_statement_line_preparer","access.bank.acc.rec.statement.line.preparer","model_bank_acc_rec_statement_line","group_bank_stmt_preparer",1,1,1,1
+"access_bank_acc_rec_statement_line_verifier","access.bank.acc.rec.statement.line.verifier","model_bank_acc_rec_statement_line","group_bank_stmt_verifier",1,1,1,1
```
Comment (bwrsandman) [inline account_banking_reconciliation/security/ir.model.access.csv:5]: Not blocking, but the quotes are redundant

File: account_banking_reconciliation/account_banking_reconciliation.py (hunk 1 above)
Comment (bwrsandman) [inline account_banking_reconciliation/account_banking_reconciliation.py]: Good

**Comment by bwrsandman** [comment]
I only have style issues.

**Comment by llacroix** [comment]
ok good.

**Comment by llacroix** [comment]
@bwrsandman I noticed there is a ChangeLog file, is it worth keeping it? There is almost nothing useful there, I guess the git log will be worth more than this txt file.

File: account_banking_make_deposit/demo/account_banking_make_deposit_demo.xml (hunk 5 above)
Comment (llacroix) [inline account_banking_make_deposit/demo/account_banking_make_deposit_demo.xml]: Done

File: account_banking_make_deposit/wizard/add_deposit_items_view.xml (hunk 7 above)
Comment (llacroix) [inline account_banking_make_deposit/wizard/add_deposit_items_view.xml:1]: Done 2 spaces on each xml files to be certain

**Comment by bwrsandman** [comment]
@llacroix take it out

**Comment by bwrsandman** [comment]
:+1:

File: account_banking_make_deposit/models/deposit_ticket.py [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+                    cr, uid, [group_verifier_id], ['res_id']
+                )[0]['res_id']
+
+                group_verifier = res_groups_obj.browse(
+                    cr, uid, res_id, context=context
+                )
+                group_user_ids = [user.id for user in group_verifier.users]
+                if deposit.state != 'draft' and uid not in group_user_ids:
+                    raise orm.except_orm(
+                        _('User Error !'),
```
Comment (gurneyalex) [inline account_banking_make_deposit/models/deposit_ticket.py]: remove exclamation mark

File: account_banking_make_deposit/models/deposit_ticket.py [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+        3.The account.move.lines are updated and written with
+          the 'Deposit Ticket #'
+        4.The status field is updated to "Done"
+        5.New GL entries are made.
+        """
+        move_lines = []
+        for deposit in self.browse(cr, uid, ids, context=context):
+            if not deposit.journal_id.sequence_id:
+                raise orm.except_orm(
+                    _('Error !'),
```
Comment (gurneyalex) [inline account_banking_make_deposit/models/deposit_ticket.py]: remove exclamation mark

File: account_banking_make_deposit/models/deposit_ticket.py [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+                    _('Please define sequence on deposit journal')
+                )
+            if deposit.journal_id.centralisation:
+                raise orm.except_orm(
+                    _('UserError'),
```
Comment (gurneyalex) [inline account_banking_make_deposit/models/deposit_ticket.py]: s/UserError/User Error/

File: account_banking_make_deposit/models/deposit_ticket_line.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    def create(self, cr, uid, vals, context=None):
+        # Any Line cannot be manually added. Use the wizard to add lines.
+        if not vals.get('move_line_id', False):
+            raise orm.except_orm(
+                _('Processing Error'),
+                _(
+                    'You cannot add any new deposit ticket line '
+                    'manually as of this revision! '
+                    'Please use the button \"Add Deposit '
+                    'Items\" to add deposit ticket line!'
```
Comment (gurneyalex) [inline account_banking_make_deposit/models/deposit_ticket_line.py]: backslash could be removed

**Comment by gurneyalex** [comment]
needs rebasing

**Comment by max3903** [comment]
- [x] Accountant group should inherit *preparer groups.
- [x] Financial manager group should inherit *verifier groups.
- [x] Fix tooltip of the period field (Field is mandatory and tooltip says it can be empty)

**Comment by max3903** [comment]
When I click on "Add deposit items" or "Save", I get:

```
Server Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/addons/web/session.py", line 89, in send
    return openerp.netsvc.dispatch_rpc(service_name, method, args)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/netsvc.py", line 300, in dispatch_rpc
    result = ExportService.getService(service_name).dispatch(method, params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/service/web_services.py", line 632, in dispatch
    res = fn(db, uid, *params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/osv/osv.py", line 193, in execute_kw
    return self.execute(db, uid, obj, method, *args, **kw or {})
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/osv/osv.py", line 133, in wrapper
    return f(self, dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/osv/osv.py", line 202, in execute
    res = self.execute_cr(cr, uid, obj, method, *args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/osv/osv.py", line 190, in execute_cr
    return getattr(object, method)(cr, uid, *args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/osv/orm.py", line 4557, in create
    self._check_selection_field_value(cr, user, field, vals[field], context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2949002-144-07d785/openerp/osv/orm.py", line 2908, in _check_selection_field_value
    elif val in dict(self._columns[field].selection(self, cr, uid, context=context)):
TypeError: <lambda>() got an unexpected keyword argument 'context'
```

**Comment by max3903** [comment]
:+1:

**Comment by veloutin** [comment]
:+1:

**Pull request merged**